File: backend-java/src/main/java/com/example/calorie/meal/dto/MealDtos.java
package com.example.calorie.meal.dto;

import com.example.calorie.common.Nutrition;
import com.example.calorie.meal.MealLog;
import com.example.calorie.meal.MealLogItem;
import com.example.calorie.meal.MealType;
import jakarta.validation.Valid;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

public final class MealDtos {

    private MealDtos() {
    }

    /**
     * 記録する 1 品。
     *
     * <p>栄養値を受け取らないのは意図的。クライアントが計算して送る形にすると、
     * Web とモバイルで計算式がずれたときに、記録されるデータそのものが食い違う。
     */
    public record MealLogItemRequest(
            @NotNull Long foodId,
            @NotNull @Positive @DecimalMax("10000") BigDecimal amountG
    ) {
    }

    public record MealLogRequest(
            @NotNull LocalDate eatenOn,
            @NotNull MealType mealType,
            @Size(max = 500) String note,
            @NotEmpty @Size(max = 50) @Valid List<MealLogItemRequest> items
    ) {
    }

    public record MealLogItemResponse(
            Long id,
            Long foodId,
            String foodName,
            BigDecimal amountG,
            Nutrition nutrition
    ) {
        public static MealLogItemResponse from(MealLogItem item) {
            return new MealLogItemResponse(
                    item.getId(), item.getFoodId(), item.getFoodName(),
                    item.getAmountG(), item.nutrition());
        }
    }

    public record MealLogResponse(
            Long id,
            LocalDate eatenOn,
            MealType mealType,
            String note,
            List<MealLogItemResponse> items,
            Nutrition total
    ) {
        public static MealLogResponse from(MealLog log) {
            return new MealLogResponse(
                    log.getId(),
                    log.getEatenOn(),
                    log.getMealType(),
                    log.getNote(),
                    log.getItems().stream().map(MealLogItemResponse::from).toList(),
                    log.total()
            );
        }
    }
}

File: backend-kotlin/src/main/kotlin/com/example/calorie/meal/dto/MealDtos.kt
package com.example.calorie.meal.dto

import com.example.calorie.common.Nutrition
import com.example.calorie.meal.MealLog
import com.example.calorie.meal.MealLogItem
import com.example.calorie.meal.MealType
import jakarta.validation.Valid
import jakarta.validation.constraints.DecimalMax
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.Positive
import jakarta.validation.constraints.Size
import java.math.BigDecimal
import java.time.LocalDate

/**
 * 記録する 1 品。
 *
 * 栄養値を受け取らないのは意図的。クライアントが計算して送る形にすると、
 * Web とモバイルで計算式がずれたときに、記録されるデータそのものが食い違う。
 */
data class MealLogItemRequest(
    val foodId: Long,
    @field:Positive @field:DecimalMax("10000") val amountG: BigDecimal,
)

data class MealLogRequest(
    val eatenOn: LocalDate,
    val mealType: MealType,
    @field:Size(max = 500) val note: String? = null,
    @field:NotEmpty @field:Size(max = 50) @field:Valid val items: List<MealLogItemRequest>,
)

data class MealLogItemResponse(
    val id: Long,
    val foodId: Long?,
    val foodName: String,
    val amountG: BigDecimal,
    val nutrition: Nutrition,
) {
    companion object {
        fun from(item: MealLogItem) = MealLogItemResponse(
            id = item.id!!,
            foodId = item.foodId,
            foodName = item.foodName,
            amountG = item.amountG,
            nutrition = item.nutrition(),
        )
    }
}

data class MealLogResponse(
    val id: Long,
    val eatenOn: LocalDate,
    val mealType: MealType,
    val note: String?,
    val items: List<MealLogItemResponse>,
    val total: Nutrition,
) {
    companion object {
        fun from(log: MealLog) = MealLogResponse(
            id = log.id!!,
            eatenOn = log.eatenOn,
            mealType = log.mealType,
            note = log.note,
            items = log.items.map(MealLogItemResponse::from),
            total = log.total(),
        )
    }
}
